### PR TITLE
feat[logging]: Replacement API for get_correlation_ids

### PR DESCRIPTION
**Pull request by Yun-Kim**
## Description
In #2357, helpers.py was deprecated which contained helper function `get_correlation_ids()` which is used by users who have custom logging (or don't use the standard logging module). Though not used anywhere in our library, `get_correlation_ids()` appears to still be used by users. 

Thus, the solution is to provide a replacement API for `get_correlation_ids()` that more closely matches the tracer's logging integration, in the sense that it now returns a dictionary which includes not only the `span_id` and `trace_id` of the current active span, but also the configuration's service, env, and version.

### CODE REVIEW

**Review by jd — APPROVED** [review]
Do we want to add some documentation around this in the user doc?

**Review by brettlangdon — CHANGES_REQUESTED** [review]
user facing docs 👍🏻 

can we add some tests that use this log record in how we expect users to use it for logging?

e.g. format strings vs "extra" vs json logging, etc

**Comment by Yun-Kim** [comment]
> user facing docs 👍🏻
> 
> can we add some tests that use this log record in how we expect users to use it for logging?
> 
> e.g. format strings vs "extra" vs json logging, etc

Does it make more sense to move this API from `trace_utils` and into the logging integration module?  I'm thinking of adding the non-standard library logging instructions (for users that need this helper) into the "Logs Injection" section in the logging integration docs, in that case it would make more sense for the helper to be in the logging module rather than `trace_utils`.

File: ddtrace/contrib/trace_utils.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 from typing import Tuple
 
 from ddtrace import Pin
+from ddtrace import Tracer
```
Comment (brettlangdon) [inline ddtrace/contrib/trace_utils.py:16]: why this change?

File: ddtrace/tracer.py [hunk 2]
```diff
@@ -246,6 +248,25 @@ def current_trace_context(self, *args, **kwargs):
             return active.context
         return None
 
+    def get_log_correlation_context(self):
+        # type: () -> Optional[DDLogRecord]
```
Comment (brettlangdon) [inline ddtrace/tracer.py:252]: should this be `Optional`? probably better if we always return a `DDLogRecord`, just whether we have active values or not.

e.g. if we just did:

```python
span = None
if self.enabled:
    span = self.current_span()

return DDLogRecord(... what you have...)
```

could be a way to get rid of the `Optional` part.

What do you think?

File: tests/tracer/test_correlation_log_context.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    def test_get_log_correlation_context_no_active_span(self):
+        """Ensure empty DDLogRecord generated if no active span."""
+        dd_log_record = tracer.get_log_correlation_context()
+        assert dd_log_record.span_id == "0"
+        assert dd_log_record.trace_id == "0"
+        assert dd_log_record.service == ""
+        assert dd_log_record.env == ""
+        assert dd_log_record.version == ""
+
+    def test_get_log_correlation_context_disabled_tracer(self):
```
Comment (brettlangdon) [inline tests/tracer/test_correlation_log_context.py:75]: I think this should be the same as the previous test, we get a record and everything is zeroed out

File: tests/tracer/test_correlation_log_context.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+def tracer_injection(logger, log_method, event_dict):
+    """Inject tracer information into custom structlog log"""
+    correlation_log_context = tracer.get_log_correlation_context()
+
+    # add ids and configs to structlog event dictionary
+    event_dict["dd.trace_id"] = correlation_log_context.trace_id
+    event_dict["dd.span_id"] = correlation_log_context.span_id
+    event_dict["dd.env"] = correlation_log_context.env
+    event_dict["dd.service"] = correlation_log_context.service
+    event_dict["dd.version"] = correlation_log_context.version
```
Comment (brettlangdon) [inline tests/tracer/test_correlation_log_context.py:29]: `event_dict.update(correlation_log_context)` could be nice to have.

that way if we add more properties to the log correlation then you'll get it automatically.

what do you think?

Comment (Yun-Kim) [reply tests/tracer/test_correlation_log_context.py:29]: Yeah I'm fine with that, I just followed what we recommend to customers on our docs: https://docs.datadoghq.com/tracing/connect_logs_and_traces/python/#no-standard-library-logging

File: ddtrace/contrib/trace_utils.py (hunk 1 above)
Comment (Yun-Kim) [reply ddtrace/contrib/trace_utils.py:16]: Haha missed this! TY for the catch

File: tests/tracer/test_correlation_log_context.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    yield
+    config.service = config.env = config.version = None
+
+
+def tracer_injection(logger, log_method, event_dict):
+    """Inject tracer information into custom structlog log"""
+    correlation_log_context = tracer.get_log_correlation_context()
+
+    # add ids and configs to structlog event dictionary
+    event_dict["dd_correlation_context"] = correlation_log_context
```
Comment (brettlangdon) [inline tests/tracer/test_correlation_log_context.py:25]: ah, we'll still want `event_dict["dd.trace_id"]`

we also need to be sure a `DDLogRecord` can be JSON encoded.

File: tests/tracer/test_correlation_log_context.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        assert dd_log_record.span_id == "0"
+        assert dd_log_record.trace_id == "0"
+        assert dd_log_record.service == ""
+        assert dd_log_record.env == ""
+        assert dd_log_record.version == ""
+
+    def test_custom_logging_injection(self):
+        """Ensure custom log injection via get_correlation_log_record returns proper active span information."""
+        capture_log = structlog.testing.LogCapture()
+        structlog.configure(processors=[tracer_injection, capture_log, structlog.processors.JSONRenderer()])
```
Comment (brettlangdon) [inline tests/tracer/test_correlation_log_context.py:86]: does `capture_log` need to come after `JSONRenderer`?

can we test that the JSON produced is correct?

File: tests/tracer/test_correlation_log_context.py (hunk 5 above)
Comment (Yun-Kim) [reply tests/tracer/test_correlation_log_context.py:25]: Just to clarify, when you suggested `event_dict.update(correlation_log_context)` did you mean to store the actual `DDLogRecord` in the event dictionary? Since the `DDLogRecord` object isn't a dictionary that can be used to update.

Comment (brettlangdon) [reply tests/tracer/test_correlation_log_context.py:25]: yeah, the end result needs to be the same (e.g. `event_dict["dd.trace_id"]` is the only way the correlation with the logs product will work, or `event_dict["dd"]["trace_id"]`), but I am wondering if we can make people's lives easier, instead of them needing to manually add each element to the `event_dict`


or what if they use another library that works a little differently, or.. something? just wondering if we want/need the flexibility here?

for the same reason, it might be nice if `DDLogRecord` was JSON serializable.


what do you think?

File: ddtrace/opentracer/tracer.py [hunk 7]
```diff
@@ -6,6 +8,7 @@
 from ddtrace import Span as DatadogSpan
 from ddtrace import Tracer as DatadogTracer
 from ddtrace.constants import FILTERS_KEY
+from ddtrace.contrib.logging.patch import DDLogRecord
```
Comment (Kyle-Verhoog) [inline ddtrace/opentracer/tracer.py:11]:
```suggestion
from ddtrace.contrib.logging import DDLogRecord
```

We should expose `DDLogRecord` officially if it's going to be in the public API imo

File: ddtrace/opentracer/tracer.py [hunk 8]
```diff
@@ -337,3 +340,12 @@ def extract(self, format, carrier):  # noqa: A002
         dd_span_ctx = ot_span_ctx._dd_context
         self._dd_tracer.context_provider.activate(dd_span_ctx)
         return ot_span_ctx
+
+    def get_log_correlation_context(self):
+        # type: () -> Optional[DDLogRecord]
+        """Retrieves the Correlation Identifiers for the current active ``Trace``
```
Comment (Kyle-Verhoog) [inline ddtrace/opentracer/tracer.py:346]:
```suggestion
        """Retrieves the Correlation Identifiers for the current active trace
```

we don't have a public data structure for traces

File: ddtrace/tracer.py [hunk 9]
```diff
@@ -246,6 +248,25 @@ def current_trace_context(self, *args, **kwargs):
             return active.context
         return None
 
+    def get_log_correlation_context(self):
+        # type: () -> DDLogRecord
+        """Retrieves the Correlation Identifiers for the current active ``Trace``
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:253]:
```suggestion
        """Retrieves the Correlation Identifiers for the current active trace
```

File: ddtrace/tracer.py [hunk 10]
```diff
@@ -246,6 +248,25 @@ def current_trace_context(self, *args, **kwargs):
...
+        This helper method generates a DDLogRecord for custom logging instrumentation including the trace id and
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:254]:
```suggestion
        Generates a DDLogRecord for custom logging instrumentation including the trace id and
```

File: ddtrace/tracer.py [hunk 11]
```diff
@@ -246,6 +248,25 @@ def current_trace_context(self, *args, **kwargs):
...
+        span id of the current active span, as well as the configured service, version, and environment names.
+        If there is no active span, an empty DDLogRecord will be returned.
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:256]: "empty" doesn't have any definition here - maybe something like "a DDLogRecord with all None values will be returned"?

File: ddtrace/tracer.py [hunk 12]
```diff
@@ -246,6 +248,25 @@ def current_trace_context(self, *args, **kwargs):
...
+        """
+        span = None
+        if self.enabled:
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:259]: This behaviour is a little strange IMO - mostly because the semantics of `tracer.enabled` aren't really specified. Technically the active span is still tracked and updated even if the tracer is "disabled" but since the spans won't be sent this decision makes sense :+1:

We should have a test case asserting this behaviour.

File: ddtrace/opentracer/tracer.py [hunk 13]
```diff
@@ -337,3 +339,12 @@ def extract(self, format, carrier):  # noqa: A002
         dd_span_ctx = ot_span_ctx._dd_context
         self._dd_tracer.context_provider.activate(dd_span_ctx)
         return ot_span_ctx
+
+    def get_log_correlation_context(self):
+        # type: () -> Dict[str, str]
+        """Retrieves the Correlation Identifiers for the current active trace.
+        This helper method generates a dictionary for custom logging instrumentation including the trace id and
```
Comment (Kyle-Verhoog) [inline ddtrace/opentracer/tracer.py:346]:
```suggestion
        Generates a dictionary for custom logging instrumentation including the trace id and
```

File: ddtrace/tracer.py [hunk 14]
```diff
@@ -246,6 +246,25 @@ def current_trace_context(self, *args, **kwargs):
             return active.context
         return None
 
+    def get_log_correlation_context(self):
+        # type: () -> Dict[str, str]
+        """Retrieves the Correlation Identifiers for the current active trace.
+        This helper method generates a dictionary for custom logging instrumentation including the trace id and
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:252]:
```suggestion
        Generates a dictionary for custom logging instrumentation including the trace id and
```

File: ddtrace/tracer.py [hunk 15]
```diff
@@ -246,6 +246,25 @@ def current_trace_context(self, *args, **kwargs):
...
+        span id of the current active span, as well as the configured service, version, and environment names.
+        If there is no active span, a dictionary with empty values will be returned.
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:254]:
```suggestion
        If there is no active span, a dictionary with empty string for each value will be returned.
```

File: tests/tracer/test_correlation_log_context.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+def tracer_injection(logger, log_method, event_dict):
+    """Inject tracer information into custom structlog log"""
+    correlation_log_context = tracer.get_log_correlation_context()
+    # add ids and configs to structlog event dictionary
+    event_dict["dd"] = correlation_log_context
+    return event_dict
+
+
+def assert_dd_log_record_matches_context(dd_log_record, span, service, env, version):
```
Comment (Kyle-Verhoog) [inline tests/tracer/test_correlation_log_context.py:28]: This logic just duplicates the logic done in the original method... kinda feels like we should be explicitly asserting these things

```python
assert dd_log_record == {
    "span_id": str(span1.span_id),
    "trace_id": str(span1.trace_id),
    "service": "test-service",
    "env": "test-env",
    "version": "test-version",
}
```

File: ddtrace/tracer.py [hunk 17]
```diff
@@ -246,6 +246,25 @@ def current_trace_context(self, *args, **kwargs):
             return active.context
         return None
 
+    def get_log_correlation_context(self):
+        # type: () -> Dict[str, str]
+        """Retrieves the Correlation Identifiers for the current active trace.
```
Comment (Kyle-Verhoog) [inline ddtrace/tracer.py:251]:
```suggestion
        """Return the data used to correlate a log with the current active trace.
```

File: tests/tracer/test_correlation_log_context.py (hunk 6 above)
Comment (Yun-Kim) [reply tests/tracer/test_correlation_log_context.py:93]: Nope, according to their docs `JSONRenderer` is usually the last processor in the chain, and trying `capture_log` after `JSONRenderer` leads to TypeErrors with the `event_dict` being mutated by the `JSONRenderer` and unable to support item assignment. The `JSONRenderer()` by default uses `json.dumps()` to format the string, and since our new API will always output a Dictionary with string keys and string values, we shouldn't need to worry about this not being serializable.

File: ddtrace/tracer.py (hunk 12 above)
Comment (Yun-Kim) [reply ddtrace/tracer.py:257]: Yup, added one: `test_get_log_correlation_context_disabled_tracer` 😄

**Comment by brettlangdon** [comment]
I tested this locally with `structlog`

```python
import time

import structlog

from ddtrace import tracer


def tracer_injection(logger, log_method, event_dict):
    # get correlation ids from current tracer context
    event_dict["dd"] = tracer.get_log_correlation_context()

    return event_dict


structlog.configure(processors=[tracer_injection, structlog.processors.JSONRenderer()])
log = structlog.get_logger()


for _ in range(100):
    with tracer.trace("root", service="log-test"):
        log.info("test")
        time.sleep(0.25)
```

image

image

image

File: ddtrace/opentracer/tracer.py [hunk 18]
```diff
@@ -4,6 +4,7 @@
 from typing import Optional
 from typing import Union
 
+
```
Comment (brettlangdon) [inline ddtrace/opentracer/tracer.py:7]:
```suggestion
```

**Review by brettlangdon — APPROVED** [review]

**Comment by Yun-Kim** [comment]
@brettlangdon that's awesome! I'm concerned the span/trace IDs don't show up though, is this expected on the logging platform?

**Comment by Yun-Kim** [comment]
The corresponding docs change should be merged after this gets merged: https://github.com/DataDog/documentation/pull/11458

**Comment by brettlangdon** [comment]
> @brettlangdon that's awesome! I'm concerned the span/trace IDs don't show up though, is this expected on the logging platform?

This is correct. They remove it from the metadata, but we can see the linked trace which means it is working correctly.

**Review by nizox — APPROVED** [review]

**Comment by brettlangdon** [comment]
@Mergifyio backport 0.52
